hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -246,6 +259,47 @@ public PrecompileContractResult computePrecompile(
                 : PrecompiledContract.PrecompileContractResult.success(result);
     }
 
+    public boolean unqualifiedDelegateDetected(MessageFrame frame) {
+        // if the first message frame is not a delegate, it's not a delegate
+        if (!isDelegateCall(frame)) {
+            return false;
+        }
+
+        final var recipient = frame.getRecipientAddress();
+        // but we accept delegates iff the token redirect contract calls us,
+        // so if they are not a token, or on the permitted callers list, then
+        // we are a delegate and we are done.
+        if (isToken(frame, recipient)
+                || dynamicProperties.permittedDelegateCallers().contains(recipient)) {
+            // make sure we have a parent calling context
+            final var stack = frame.getMessageFrameStack();
+            final var frames = stack.iterator();
+            frames.next();
+            if (!frames.hasNext()) {
+                // Impossible to get here w/o a catastrophic EVM bug
+                log.error("Delegate call frame had no parent");
+                return false;
```

Comment:
Should we set this to true? Or throw something?

Reply:
Logged `Possibly CATASTROPHIC` instead; note this code _is_ unreachable since a top-level EVM transaction must necessarily have the same `receiver` and `contract` addresses.